File: test/rule/rule-spec.js
'use strict'

const chai = require('chai')

const mockMap = require('../mock/map')
const MockOrder = require('../mock/order')
const MockRule = require('./../mock/rule')

const { Unit, Board } = require('./../../lib/board/package')
const ResolvedResult = require('./../../lib/rule/resolved-result')
const { Executed, Replaced } = require('./../../lib/rule/order-result')

chai.should()

describe('A rule', () => {
  const [Army, Fleet] = mockMap.militaryBranches
  const [spain, naples, apulia, western] = mockMap.provinces
  const [spa, spaSc, nap, apu, wes] = mockMap.locations
  const map = mockMap.map
  const unit = new Unit(Fleet, nap, 'France')
  const board = new Board(map, 'State', [unit], [], [])

  describe('when one unit has several orders', () => {
    it('does not resolve orders.', () => {
      const rule = new MockRule()
      rule.resolve(board, [new MockOrder(unit), new MockOrder(unit)]).should.deep.equal({
        err: 'F Nap: several orders'
      })
    })
  })
  describe('when resolving an invalid order', () => {
    it('uses a default order', () => {
      const unit1 = new Unit(Fleet, nap, 'France')
      const unit2 = new Unit(Army, spa, 'France')
      const board = new Board(null, 'State', [unit1, unit2], [], [])

      const rule = new MockRule()
      rule.errorMessageOfOrder = (board, order) => {
        return (order.unit === unit1 || order.replaced) ? null : 'Invalid'
      }
      rule.defaultOrderOf = (board, unit) => {
        const o = new MockOrder(unit)
        o.replaced = true
        return o
      }
      rule.resolveProcedure = (board, orders) => {
        return {
          result: new ResolvedResult(
            board,
            orders.map(order => new Executed(order, 'Resolved')),
            true
          )
        }
      }

      const results = rule.resolve(board, [new MockOrder(unit1), new MockOrder(unit2)]).result.results

      const tmp = new MockOrder(unit2)
      tmp.replaced = true
      results.should.deep.equal(new Set([
        new Executed(new MockOrder(unit1), 'Resolved'),
        new Replaced(new MockOrder(unit2), 'Invalid', tmp, 'Resolved')
      ]))
    })
  })
  describe('when the set of orders are invalid', () => {
    it('does not resolve the orders', () => {
      const rule = new MockRule()
      rule.errorMessageOfOrders = (board, orders) => 'Invalid'
      rule.resolve(board, [new MockOrder(unit)]).should.deep.equal({
        err: 'Invalid'
      })
    })
  })
  describe('when several unit that require an order do not have orders', () => {
    it('uses a default order.', () => {
      const unit1 = new Unit(Fleet, nap, 'France')
      const unit2 = new Unit(Army, spa, 'France')
      const board = new Board(null, 'State', [unit1, unit2], [], [])

      const rule = new MockRule()
      rule.unitsRequiringOrder = (board) => board.units
      rule.defaultOrderOf = (board, unit) => {
        const o = new MockOrder(unit)
        return o
      }
      rule.resolveProcedure = (board, orders) => {
        return {
          result: new ResolvedResult(
            board,
            orders.map(order => new Executed(order, 'Resolved')),
            true
          )
        }
      }
      const result = rule.resolve(board, []).result.results
      result.should.deep.equal(new Set([
        new Executed(new MockOrder(unit1), 'Resolved'),
        new Executed(new MockOrder(unit2), 'Resolved')
      ]))
    })
  })
})

File: core/shared/src/test/scala/diplomacy/rule/RuleSpec.scala
package diplomacy.rule

import diplomacy.UnitSpec
import diplomacy.board.{ Board => BaseBoard, BoardHelper }
import diplomacy.mock.board._
import diplomacy.mock.rule._

class RuleSpec extends UnitSpec {
  "A rule" when {
    "one unit has several orders" should {
      "not resolve the orders." in {
        val board = new Board(
          map, "State", Set(new DiplomacyUnit(france, army, mar)),
          Map(), Map()
        )
        val $ = BoardHelper(board)

        val rule = new MockRule() {}
        rule.resolve(board, Set(Move($.A($.Mar), $.Spa), Hold($.A($.Mar)))) should be(
          Left(InvalidOrderMessage("A Mar: several orders."))
        )
      }
    }
    "resolving an invalid order" should {
      "uses a default order." in {
        val board = new Board(
          map, "State",
          Set(new DiplomacyUnit(france, army, mar), new DiplomacyUnit(france, fleet, wes)),
          Map(), Map()
        )
        val $ = BoardHelper(board)

        val rule = new MockRule() {
          override def errorMessageOfOrder(board: Board)(order: Order) = order match {
            case _: Hold => None
            case _ => Option(InvalidOrderMessage(""))
          }
          override def defaultOrderOf(board: Board)(unit: DiplomacyUnit) = Hold(unit)
          override def resolveProcedure(board: Board, orders: Set[Order]) = {
            Right(new ResolvedResult(
              board,
              orders map { order =>
                OrderResult.Executed[Power, MilitaryBranch, Order, String](order, "resolved")
              },
              false
            ))
          }
        }

        val Right(result) = rule.resolve(board, Set(Move($.A($.Mar), $.Spa), Hold($.F($.Wes))))
        result.result should be (Set(
          OrderResult.Executed[Power, MilitaryBranch, rule.Order, String](
            Hold($.F($.Wes)), "resolved"
          ),
          OrderResult.Replaced[Power, MilitaryBranch, rule.Order, String](
            Move($.A($.Mar), $.Spa), InvalidOrderMessage(""), Hold($.A($.Mar)), "resolved"
          )
        ))
      }
    }
    "the set or orders are invalid" should {
      "not resolve the orders." in {
        val board = new Board(
          map, "State",
          Set(new DiplomacyUnit(france, army, mar), new DiplomacyUnit(france, fleet, wes)),
          Map(), Map()
        )

        val rule = new MockRule() {
          override def errorMessageOfOrders(board: Board)(orders: Set[Order]) = {
            Option(InvalidOrderMessage(""))
          }
        }
        rule.resolve(board, Set()) should be(Left(InvalidOrderMessage("")))
      }
    }
    "several unit that requre an order do not have orders" should {
      "uses a default order." in {
        val board = new Board(
          map, "State",
          Set(new DiplomacyUnit(france, army, mar), new DiplomacyUnit(france, fleet, wes)),
          Map(), Map()
        )
        val $ = BoardHelper(board)

        val rule = new MockRule() {
          override def defaultOrderOf(board: Board)(unit: DiplomacyUnit) = Hold(unit)
          override def resolveProcedure(board: Board, orders: Set[Order]) = {
            Right(new ResolvedResult(
              board,
              orders map { order =>
                OrderResult.Executed[Power, MilitaryBranch, Order, String](order, "resolved")
              },
              false
            ))
          }
          override def unitsRequiringOrder(board: Board) = Set($.A($.Mar))
        }

        val Right(result) = rule.resolve(board, Set())
        result.result should be (Set(
          OrderResult.Executed[Power, MilitaryBranch, rule.Order, String](
            Hold($.A($.Mar)), "resolved"
          )
        ))
      }
    }
  }
}
